### DISCUSSION TITLE
How to avoid this error in Windows? ImportError: cannot import name 'RRef' from 'torch.distributed.rpc'

### DISCUSSION THREAD

**Discussion by yiluzhou1** [Project-MONAI/MONAILabel#1465] [Q&A] [answered]
I'm using Windows 10 laptop with RTX A1000 GPU. I installed MonaLabel (with CUDA). However, the following error popped out: 
 ImportError: cannot import name 'RRef' from 'torch.distributed.rpc'
It seems it is related to unavailability of rpc in Pytorch Windows. How to bypass this issue? Thank you!

The entire log is as below:

> C:\Development\monailabel>monailabel start_server --app apps/radiology --studies datasets/Task09_Spleen/imagesTr --conf models segmentation
> Using PYTHONPATH=C:\Users\yzhou\AppData\Local\miniconda3\envs;C:\Users\yzhou\AppData\Local\miniconda3\envs;
> ""
> [2023-06-08 16:45:15,172] [5692] [MainThread] [INFO] (__main__:285) - USING:: version = False
> [2023-06-08 16:45:15,172] [5692] [MainThread] [INFO] (__main__:285) - USING:: app = C:\Development\monailabel\apps\radiology
> [2023-06-08 16:45:15,173] [5692] [MainThread] [INFO] (__main__:285) - USING:: studies = C:\Development\monailabel\datasets\Task09_Spleen\imagesTr
> [2023-06-08 16:45:15,174] [5692] [MainThread] [INFO] (__main__:285) - USING:: verbose = INFO
> [2023-06-08 16:45:15,174] [5692] [MainThread] [INFO] (__main__:285) - USING:: conf = [['models', 'segmentation']]
> [2023-06-08 16:45:15,174] [5692] [MainThread] [INFO] (__main__:285) - USING:: host = 0.0.0.0
> [2023-06-08 16:45:15,174] [5692] [MainThread] [INFO] (__main__:285) - USING:: port = 8000
> [2023-06-08 16:45:15,174] [5692] [MainThread] [INFO] (__main__:285) - USING:: uvicorn_app = monailabel.app:app
> [2023-06-08 16:45:15,175] [5692] [MainThread] [INFO] (__main__:285) - USING:: ssl_keyfile = None
> [2023-06-08 16:45:15,175] [5692] [MainThread] [INFO] (__main__:285) - USING:: ssl_certfile = None
> [2023-06-08 16:45:15,175] [5692] [MainThread] [INFO] (__main__:285) - USING:: ssl_keyfile_password = None
> [2023-06-08 16:45:15,175] [5692] [MainThread] [INFO] (__main__:285) - USING:: ssl_ca_certs = None
> [2023-06-08 16:45:15,175] [5692] [MainThread] [INFO] (__main__:285) - USING:: workers = None
> [2023-06-08 16:45:15,175] [5692] [MainThread] [INFO] (__main__:285) - USING:: limit_concurrency = None
> [2023-06-08 16:45:15,176] [5692] [MainThread] [INFO] (__main__:285) - USING:: access_log = False
> [2023-06-08 16:45:15,176] [5692] [MainThread] [INFO] (__main__:285) - USING:: root_path = /
> [2023-06-08 16:45:15,176] [5692] [MainThread] [INFO] (__main__:285) - USING:: log_level = info
> [2023-06-08 16:45:15,176] [5692] [MainThread] [INFO] (__main__:285) - USING:: log_config = None
> [2023-06-08 16:45:15,176] [5692] [MainThread] [INFO] (__main__:285) - USING:: dryrun = False
> [2023-06-08 16:45:15,176] [5692] [MainThread] [INFO] (__main__:285) - USING:: action = start_server
> [2023-06-08 16:45:15,177] [5692] [MainThread] [INFO] (__main__:296) -
> Allow Origins: ['*']
> [2023-06-08 16:45:16,657] [5692] [MainThread] [INFO] (uvicorn.error:74) - Started server process [5692]
> [2023-06-08 16:45:16,657] [5692] [MainThread] [INFO] (uvicorn.error:48) - Waiting for application startup.
> [2023-06-08 16:45:16,657] [5692] [MainThread] [INFO] (monailabel.interfaces.utils.app:37) - Initializing App from: C:\Development\monailabel\apps\radiology; studies: C:\Development\monailabel\datasets\Task09_Spleen\imagesTr; conf: {'models': 'segmentation'}
> [2023-06-08 16:45:17,863] [5692] [MainThread] [INFO] (monailabel.utils.others.class_utils:57) - Subclass for MONAILabelApp Found: <class 'main.MyApp'>
> [2023-06-08 16:45:17,878] [5692] [MainThread] [ERROR] (uvicorn.error:123) - Traceback (most recent call last):
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\starlette\routing.py", line 677, in lifespan
>     async with self.lifespan_context(app) as maybe_state:
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\starlette\routing.py", line 566, in __aenter__
>     await self._router.startup()
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\starlette\routing.py", line 654, in startup
>     await handler()
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\monailabel\app.py", line 108, in startup_event
>     instance = app_instance()
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\monailabel\interfaces\utils\app.py", line 50, in app_instance
>     app = c(app_dir=app_dir, studies=studies, conf=conf)
>   File "C:\Development\monailabel\apps\radiology\main.py", line 49, in __init__
>     for c in get_class_names(lib.configs, "TaskConfig"):
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\monailabel\utils\others\class_utils.py", line 164, in get_class_names
>     module = importlib.import_module("." + name, package=current_module_name)
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\importlib\__init__.py", line 127, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 850, in exec_module
>   File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
>   File "C:\Development\monailabel\apps\radiology\lib\configs\deepedit.py", line 17, in <module>
>     import lib.trainers
>   File "C:\Development\monailabel\apps\radiology\lib\trainers\__init__.py", line 12, in <module>
>     from .deepedit import DeepEdit
>   File "C:\Development\monailabel\apps\radiology\lib\trainers\deepedit.py", line 45, in <module>
>     from monailabel.tasks.train.basic_train import BasicTrainTask, Context
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\monailabel\tasks\train\basic_train.py", line 23, in <module>
>     import ignite
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\ignite\__init__.py", line 3, in <module>
>     import ignite.engine
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\ignite\engine\__init__.py", line 10, in <module>
>     from ignite.metrics import Metric
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\ignite\metrics\__init__.py", line 7, in <module>
>     from ignite.metrics.frequency import Frequency
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\ignite\metrics\frequency.py", line 7, in <module>
>     from ignite.handlers.timing import Timer
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\ignite\handlers\__init__.py", line 5, in <module>
>     from ignite.handlers.checkpoint import Checkpoint, DiskSaver, ModelCheckpoint
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\ignite\handlers\checkpoint.py", line 17, in <module>
>     from torch.distributed.optim import ZeroRedundancyOptimizer
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\torch\distributed\optim\__init__.py", line 8, in <module>
>     from .optimizer import DistributedOptimizer
>   File "C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\torch\distributed\optim\optimizer.py", line 9, in <module>
>     from torch.distributed.rpc import RRef
> ImportError: cannot import name 'RRef' from 'torch.distributed.rpc' (C:\Users\yzhou\AppData\Local\miniconda3\envs\TEST\lib\site-packages\torch\distributed\rpc\__init__.py)
> 
> [2023-06-08 16:45:17,879] [5692] [MainThread] [ERROR] (uvicorn.error:59) - Application startup failed. Exiting.

**Comment by tangy5** [comment, marked as answer]
Emmm, it looks like the PyTorch Distributed package does not support Windows well on your machine. You could refer to Pytorch installation guide. 

Or you could try use the monailabel docker, maybe a better choice to make it compatible with your machine without bothering too many packages installation issues. 

`docker run -it --rm --gpus all --ipc=host --net=host -v ~:/workspace/ projectmonai/monailabel:latest bash
`

**Reply by yiluzhou1** [reply]
@tangy5 Thank you so much for your advice!
